src/braket/circuits/circuit.py

```diff
@@ -1043,25 +1050,110 @@ def diagram(self, circuit_diagram_class=AsciiCircuitDiagram) -> str:
+    ) -> Union[OpenQasmProgram, JaqcdProgram]:
         """
         Converts the circuit into the canonical intermediate representation.
         If the circuit is sent over the wire, this method is called before it is sent.
 
+        Args:
+            ir_type (IRType): The IRType to use for converting the circuit object to its
+                IR representation.
+            serialization_properties (SerializationProperties): The serialization properties to use
+                while serializing the object to the IR representation. The serialization properties
+                supplied must correspond to the supplied `ir_type`. Defaults to None.
+
         Returns:
-            (Program): An AWS quantum circuit description program in JSON format.
+            (Union[OpenQasmProgram, JaqcdProgram]): A representation of the circuit in the
+            `ir_type` format.
+
+        Raises:
+            ValueError: If the supplied `ir_type` is not supported, or if the supplied serialization
+            properties don't correspond to the `ir_type`.
         """
-        ir_instructions = [instr.to_ir() for instr in self.instructions]
-        ir_results = [result_type.to_ir() for result_type in self.result_types]
+        if ir_type == IRType.JAQCD:
+            return self._to_jaqcd()
+        elif ir_type == IRType.OPENQASM:
+            if serialization_properties and not isinstance(
+                serialization_properties, OpenQASMSerializationProperties
+            ):
+                raise ValueError(
+                    "serialization_properties must be of type OpenQASMSerializationProperties "
+                    "for IRType.OPENQASM."
+                )
+            return self._to_openqasm(serialization_properties or OpenQASMSerializationProperties())
+        else:
+            raise ValueError(f"Supplied ir_type {ir_type} is not supported.")
+
+    def _to_jaqcd(self) -> JaqcdProgram:
+        jaqcd_ir_type = IRType.JAQCD
+        ir_instructions = [instr.to_ir(ir_type=jaqcd_ir_type) for instr in self.instructions]
+        ir_results = [result_type.to_ir(ir_type=jaqcd_ir_type) for result_type in self.result_types]
         ir_basis_rotation_instructions = [
-            instr.to_ir() for instr in self.basis_rotation_instructions
+            instr.to_ir(ir_type=jaqcd_ir_type) for instr in self.basis_rotation_instructions
         ]
-        return Program.construct(
+        return JaqcdProgram.construct(
             instructions=ir_instructions,
             results=ir_results,
             basis_rotation_instructions=ir_basis_rotation_instructions,
         )
 
+    def _to_openqasm(
+        self, serialization_properties: OpenQASMSerializationProperties
+    ) -> OpenQasmProgram:
+        ir_instructions, qubit_reference_format = self._create_openqasm_header(
+            serialization_properties
+        )
+        serialization_properties.qubit_reference_format = qubit_reference_format
```

Comment:
What if this was already set by whoever created the properties? Do we get to overwrite it?

It seems like we should be asking the properties for this information, rather than setting it?

Reply:
It doesn't make sense for the user to set it and would be ignored. Removed.